cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java

```diff
@@ -112,9 +119,88 @@ private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties
     return table;
   }
 
+  private MetricsTable getOrCreateResolutionMetricsTable(String v2TableName, String v3TableName,
+                                                 TableProperties.Builder props) {
+    try {
+      // metrics tables are in the system namespace
+      DatasetId v2TableId = NamespaceId.SYSTEM.dataset(v2TableName);
+
+      MetricsTable v2Table = null;
+      if (dsFramework.hasInstance(v2TableId)) {
+        v2Table = DatasetsUtil.getOrCreateDataset(dsFramework, v2TableId, MetricsTable.class.getName(),
```

Comment:
Yes, but you are creating a race condition here. If the migration finishes, it will delete the v2 Table. That can happen in between the hasInstance() and the getOrCreate(). This would then recreate the v2Table. It is better to simply use datasetFramework.getDataset(). If the table exists, it will return non-null.